File: day10/main.cpp
// https://adventofcode.com/2024/day/10

#include <fstream>
#include <iostream>
#include <iterator>
#include <set>
#include <sstream>
#include <vector>

using CoordinateType = std::pair<int, int>;

std::vector<CoordinateType>
findTrailheadsPos(const std::vector<std::string> &topographyMap) {
  std::vector<CoordinateType> trailheads;
  for (int i{}; i < topographyMap.size(); i++)
    for (int j{}; j < topographyMap[i].size(); j++)
      trailheads.push_back({i, j});

  return trailheads;
}

bool inBounds(const std::vector<std::string> &topographyMap, int x, int y) {
  return x >= 0 && y >= 0 && x < topographyMap.size() &&
         y < topographyMap[x].size();
}

int searchHelper(const std::vector<std::string> &topographyMap, int x, int y,
                 int expectedValue, std::set<CoordinateType> *seen) {

  if (!inBounds(topographyMap, x, y))
    return 0;

  int currentValue = topographyMap[x][y] - '0';

  if (currentValue != expectedValue)
    return 0;

  if (seen && seen->count({x, y}))
    return 0;

  if (expectedValue == 9) {
    if (seen)
      seen->insert({x, y});
    return 1;
  }

  int nextValue = expectedValue + 1;

  int right = searchHelper(topographyMap, x + 1, y, nextValue, seen);
  int left = searchHelper(topographyMap, x - 1, y, nextValue, seen);
  int down = searchHelper(topographyMap, x, y + 1, nextValue, seen);
  int up = searchHelper(topographyMap, x, y - 1, nextValue, seen);

  int totalPaths = right + up + left + down;

  return totalPaths;
}

int searchTrail(const std::vector<std::string> &topographyMap, int x, int y,
                int expectedValue, std::set<CoordinateType> &seen) {

  return searchHelper(topographyMap, x, y, expectedValue, &seen);
}

int searchAllTrails(const std::vector<std::string> &topographyMap, int x, int y,
                    int expectedValue) {

  return searchHelper(topographyMap, x, y, expectedValue, nullptr);
}

std::string getRawText(const std::string &filename) {

  std::ifstream f(filename);
  if (!f.is_open())
    throw std::runtime_error("Error opening file");

  std::stringstream buffer;
  buffer << f.rdbuf();
  return buffer.str();
}

std::vector<std::string> extractLines(const std::string &text) {
  std::istringstream iss(text);
  return {std::istream_iterator<std::string>(iss),
          std::istream_iterator<std::string>()};
}

int main(int argc, char **argv) {

  const char *filename = argc > 1 ? argv[1] : "input.txt";

  std::string fileContent = getRawText(filename);

  std::vector<std::string> topographyMap = extractLines(fileContent);

  std::vector<CoordinateType> trailheads = findTrailheadsPos(topographyMap);

  int partialTrailCount{}; // Didn't know how else to call, because I did part2
                           // first somehow
  int totalTrailCount{};

  for (auto &startingPos : trailheads) {
    std::set<CoordinateType> seen;
    int partialTrail = searchTrail(topographyMap, startingPos.first,
                                   startingPos.second, 0, seen);

    int fullTrails = searchAllTrails(topographyMap, startingPos.first,
                                     startingPos.second, 0);

    partialTrailCount += partialTrail;
    totalTrailCount += fullTrails;
  }

  std::cout << partialTrailCount << std::endl;
  std::cout << totalTrailCount << std::endl;
}

File: day10/src/main.rs
use std::collections::HashSet;

type Coord = (i32, i32);

fn find_trailhead_pos(topography_map: &[String]) -> Vec<Coord> {
    let mut trailheads = Vec::new();
    for (i, row) in topography_map.iter().enumerate() {
        for j in 0..row.len() {
            trailheads.push((i as i32, j as i32));
        }
    }
    trailheads
}

fn in_bounds(topography_map: &[String], pos: Coord) -> bool {
    let (x, y) = pos;
    x >= 0
        && y >= 0
        && x < topography_map.len() as i32
        && y < topography_map[x as usize].len() as i32
}

fn search_helper(
    topography_map: &[String],
    pos: Coord,
    expected_value: i32,
    mut seen: Option<&mut HashSet<Coord>>,
) -> usize {
    if !in_bounds(topography_map, pos) {
        return 0;
    }

    let (x, y) = pos;

    let current_char = topography_map[x as usize].chars().nth(y as usize).unwrap();
    let current_value = current_char.to_digit(10).unwrap() as i32;

    if current_value != expected_value {
        return 0;
    }

    if let Some(seen) = seen.as_deref() {
        if seen.contains(&pos) {
            return 0;
        }
    }

    if expected_value == 9 {
        if let Some(seen) = seen {
            seen.insert(pos);
        }
        return 1;
    }

    let next_value = expected_value + 1;

    let right = search_helper(topography_map, (x + 1, y), next_value, seen.as_deref_mut());
    let left = search_helper(topography_map, (x - 1, y), next_value, seen.as_deref_mut());
    let down = search_helper(topography_map, (x, y + 1), next_value, seen.as_deref_mut());
    let up = search_helper(topography_map, (x, y - 1), next_value, seen);

    right + left + down + up
}

fn search_trail(
    topography_map: &[String],
    pos: Coord,
    expected_value: i32,
    seen: &mut HashSet<Coord>,
) -> usize {
    search_helper(topography_map, pos, expected_value, Some(seen))
}

fn search_all_trails(topography_map: &[String], pos: Coord, expected_value: i32) -> usize {
    search_helper(topography_map, pos, expected_value, None)
}

fn main() -> anyhow::Result<()> {
    let filename = std::env::args().nth(1).unwrap_or("input.in".to_string());

    let file_content = utility::get_raw_text(&filename)?;
    let topography_map = utility::get_vec_from_txt(&file_content);

    let trailheads = find_trailhead_pos(&topography_map);

    let mut partial_trail_count = 0;
    let mut total_trail_count = 0;

    for starting_pos in trailheads {
        let mut seen = HashSet::new();

        let partial = search_trail(&topography_map, starting_pos, 0, &mut seen);
        let full_trails = search_all_trails(&topography_map, starting_pos, 0);

        partial_trail_count += partial;
        total_trail_count += full_trails;
    }

    println!("Partial trail count: {}", partial_trail_count);
    println!("Total trail count: {}", total_trail_count);

    Ok(())
}
